mockstore/mocktikv/mvcc_leveldb.go

```diff
@@ -0,0 +1,1003 @@
+func MustNewMVCCStore() MVCCStore {
+	mvccStore, err := NewMVCCLevelDB("")
+	if err != nil {
+		panic(err)
+	}
+	return mvccStore
+}
+
+// NewMVCCLevelDB returns a new MVCCLevelDB object.
+func NewMVCCLevelDB(path string) (*MVCCLevelDB, error) {
+	var (
+		d   *leveldb.DB
+		err error
+	)
+	if path == "" {
+		d, err = leveldb.Open(storage.NewMemStorage(), nil)
+	} else {
+		d, err = leveldb.OpenFile(path, &opt.Options{BlockCacheCapacity: 600 * 1024 * 1024})
+	}
+
+	return &MVCCLevelDB{db: d}, errors.Trace(err)
+}
+
+// Iterator wraps iterator.Iterator to provide Valid() method.
+type Iterator struct {
+	iterator.Iterator
+	valid bool
+}
+
+// Next moves the iterator to the next key/value pair.
+func (iter *Iterator) Next() {
+	iter.valid = iter.Iterator.Next()
+}
+
+// Valid returns whether the iterator is exhausted.
+func (iter *Iterator) Valid() bool {
+	return iter.valid
+}
+
+func newIterator(db *leveldb.DB, slice *util.Range) *Iterator {
+	iter := &Iterator{db.NewIterator(slice, nil), true}
+	iter.Next()
+	return iter
+}
+
+func newScanIterator(db *leveldb.DB, startKey, endKey []byte) (*Iterator, []byte, error) {
+	var start, end []byte
+	if len(startKey) > 0 {
+		start = mvccEncode(startKey, lockVer)
+	}
+	if len(endKey) > 0 {
+		end = mvccEncode(endKey, lockVer)
+	}
+	iter := newIterator(db, &util.Range{
+		Start: start,
+		Limit: end,
+	})
+	// newScanIterator must handle startKey is nil, in this case, the real startKey
+	// should be change the frist key of the store.
+	if len(startKey) == 0 && iter.Valid() {
+		key, _, err := mvccDecode(iter.Key())
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		startKey = key
+	}
+	return iter, startKey, nil
+}
+
+// iterDecoder tries to decode an Iterator value.
+// If current iterator value can be decoded by this decoder, store the value and call iter.Next(),
+// Otherwise current iterator is not touched and returns false.
+type iterDecoder interface {
+	Decode(iter *Iterator) (bool, error)
+}
+
+type lockDecoder struct {
+	lock      mvccLock
+	expectKey []byte
+}
+
+// lockDecoder decodes the lock value if current iterator is at expectKey::lock.
+func (dec *lockDecoder) Decode(iter *Iterator) (bool, error) {
+	if iter.Error() != nil || !iter.Valid() {
+		return false, iter.Error()
+	}
+
+	iterKey := iter.Key()
+	key, ver, err := mvccDecode(iterKey)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if !bytes.Equal(key, dec.expectKey) {
+		return false, nil
+	}
+	if ver != lockVer {
+		return false, nil
+	}
+
+	var lock mvccLock
+	err = lock.UnmarshalBinary(iter.Value())
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	dec.lock = lock
+	iter.Next()
+	return true, nil
+}
+
+type valueDecoder struct {
+	value     mvccValue
+	expectKey []byte
+}
+
+// valueDecoder decodes a mvcc value if iter key is expectKey.
+func (dec *valueDecoder) Decode(iter *Iterator) (bool, error) {
+	if iter.Error() != nil || !iter.Valid() {
```

Comment:
It seems the validation has already been checked outside of this function except https://github.com/tikv/client-go/blob/7a23d6e7bf526a3b30847884c4005f0d3ddcc56f/mockstore/mocktikv/mvcc_leveldb.go#L545

Reply:
Copied from tidb directly. /cc @tiancaiamao Do you have any ideas?

Reply:
There is a `valid` field here  https://github.com/tikv/client-go/blob/7a23d6e7bf526a3b30847884c4005f0d3ddcc56f/mockstore/mocktikv/mvcc_leveldb.go#L125

If `iter.valid` is false, the `Decode` function should return as soon as possible. @rleungx 

The code is copied, I recommend not to modify it in this PR, if any bug is found, use a separate one to fix it.
b.t.w, tikv/client will be moved from tidb to this repo later? @disksing

Reply:
Yes. That's exactly I'm working on.